<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf8">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/3.0.0/tocbot.css">
    <!-- <link rel="stylesheet" href="docs/style.css"> -->
    <style>
      body {
  padding: 40px 80px;
  font: 14px/1.5 "Helvetica Neue", Helvetica, sans-serif;
  background: #181818 url(images/bg.png);
  text-align: center;
}

#content {
  margin: 0 auto;
  padding: 10px 40px;
  text-align: left;
  background: white;
  width: 50%;
  -webkit-border-radius: 2px;
  -moz-border-radius: 2px;
  border-radius: 2px;
  -webkit-box-shadow: 0 2px 5px 0 black;
}

#menu {
  font-size: 13px;
  margin: 0;
  padding: 0;
  text-align: left;
  position: fixed;
  top: 15px;
  left: 15px;
}

#menu ul {
  margin: 0;
  padding: 0;
}

#menu li {
  list-style: none;
}

#menu a {
  color: rgba(255,255,255,.5);
  text-decoration: none;
}

#menu a:hover {
  color: white;
}

#menu .active a {
  color: white;
}

pre {
  padding: 10px;
}

code {
  font-family: monaco, monospace, sans-serif;
  font-size: 0.85em;
}

p code {
  border: 1px solid #ECEA75;
  padding: 1px 3px;
  -webkit-border-radius: 2px;
  -moz-border-radius: 2px;
  border-radius: 2px;
  background: #FDFCD1;
}

pre {
  padding: 20px 25px;
  border: 1px solid #ddd;
  -webkit-box-shadow: inset 0 0 5px #eee;
  -moz-box-shadow: inset 0 0 5px #eee;
  box-shadow: inset 0 0 5px #eee;
}

code .comment { color: #ddd }
code .init { color: #2F6FAD }
code .string { color: #5890AD }
code .keyword { color: #8A6343 }
code .number { color: #2F6FAD }

/* override tocbot style to avoid vertical white line in table of content */
.toc-link::before {
  content: initial;
}

.error {
  color:red;
  font-weight:bold;
}

    </style>
  </head>
  <body>
    <ul id="menu"></ul>
    <div id="content">123 undefined
    <section class="suite">
      <h1>Promises/A+ Tests</h1>
      <dl>
        <section class="suite">
          <h1>2.1.2.1: When fulfilled, a promise: must not transition to any other state.</h1>
          <dl>
            <dt>already-fulfilled</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>test(resolved(value), done);</code></pre></dd>
            <dt>immediately-fulfilled</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
test(d.promise, done);
d.resolve(value);</code></pre></dd>
            <dt>eventually-fulfilled</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
test(d.promise, done);
setTimeout(function () {
    d.resolve(value);
}, 50);</code></pre></dd>
            <dt>trying to fulfill then immediately reject</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
var onFulfilledCalled = false;
d.promise.then(function onFulfilled() {
    onFulfilledCalled = true;
}, function onRejected() {
    assert.strictEqual(onFulfilledCalled, false);
    done();
});
d.resolve(dummy);
d.reject(dummy);
setTimeout(done, 100);</code></pre></dd>
            <dt>trying to fulfill then reject, delayed</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
var onFulfilledCalled = false;
d.promise.then(function onFulfilled() {
    onFulfilledCalled = true;
}, function onRejected() {
    assert.strictEqual(onFulfilledCalled, false);
    done();
});
setTimeout(function () {
    d.resolve(dummy);
    d.reject(dummy);
}, 50);
setTimeout(done, 100);</code></pre></dd>
            <dt>trying to fulfill immediately then reject delayed</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
var onFulfilledCalled = false;
d.promise.then(function onFulfilled() {
    onFulfilledCalled = true;
}, function onRejected() {
    assert.strictEqual(onFulfilledCalled, false);
    done();
});
d.resolve(dummy);
setTimeout(function () {
    d.reject(dummy);
}, 50);
setTimeout(done, 100);</code></pre></dd>
          </dl>
        </section>
        <section class="suite">
          <h1>2.1.3.1: When rejected, a promise: must not transition to any other state.</h1>
          <dl>
            <dt>already-rejected</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>test(rejected(reason), done);</code></pre></dd>
            <dt>immediately-rejected</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
test(d.promise, done);
d.reject(reason);</code></pre></dd>
            <dt>eventually-rejected</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
test(d.promise, done);
setTimeout(function () {
    d.reject(reason);
}, 50);</code></pre></dd>
            <dt>trying to reject then immediately fulfill</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
var onRejectedCalled = false;
d.promise.then(function onFulfilled() {
    assert.strictEqual(onRejectedCalled, false);
    done();
}, function onRejected() {
    onRejectedCalled = true;
});
d.reject(dummy);
d.resolve(dummy);
setTimeout(done, 100);</code></pre></dd>
            <dt>trying to reject then fulfill, delayed</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
var onRejectedCalled = false;
d.promise.then(function onFulfilled() {
    assert.strictEqual(onRejectedCalled, false);
    done();
}, function onRejected() {
    onRejectedCalled = true;
});
setTimeout(function () {
    d.reject(dummy);
    d.resolve(dummy);
}, 50);
setTimeout(done, 100);</code></pre></dd>
            <dt>trying to reject immediately then fulfill delayed</dt>
            <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
            <dd><pre><code>var d = deferred();
var onRejectedCalled = false;
d.promise.then(function onFulfilled() {
    assert.strictEqual(onRejectedCalled, false);
    done();
}, function onRejected() {
    onRejectedCalled = true;
});
d.reject(dummy);
setTimeout(function () {
    d.resolve(dummy);
}, 50);
setTimeout(done, 100);</code></pre></dd>
          </dl>
        </section>
        <section class="suite">
          <h1>2.2.1: Both &#x60;onFulfilled&#x60; and &#x60;onRejected&#x60; are optional arguments.</h1>
          <dl>
            <section class="suite">
              <h1>2.2.1.1: If &#x60;onFulfilled&#x60; is not a function, it must be ignored.</h1>
              <dl>
                <section class="suite">
                  <h1>applied to a directly-rejected promise</h1>
                  <dl>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;undefined&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;null&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;false&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;5&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is an object</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                  </dl>
                </section>
                <section class="suite">
                  <h1>applied to a promise rejected and then chained off of</h1>
                  <dl>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;undefined&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(function () { }, undefined).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;null&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(function () { }, undefined).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;false&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(function () { }, undefined).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;5&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(function () { }, undefined).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is an object</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>rejected(dummy).then(function () { }, undefined).then(nonFunction, function () {
    done();
});</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                  </dl>
                </section>
              </dl>
            </section>
            <section class="suite">
              <h1>2.2.1.2: If &#x60;onRejected&#x60; is not a function, it must be ignored.</h1>
              <dl>
                <section class="suite">
                  <h1>applied to a directly-fulfilled promise</h1>
                  <dl>
                    <dt class="error">&#x60;onRejected&#x60; is &#x60;undefined&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onRejected&#x60; is &#x60;null&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onRejected&#x60; is &#x60;false&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onRejected&#x60; is &#x60;5&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                    <dt class="error">&#x60;onRejected&#x60; is an object</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                  </dl>
                </section>
                <section class="suite">
                  <h1>applied to a promise fulfilled and then chained off of</h1>
                  <dl>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;undefined&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(undefined, function () { }).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;null&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(undefined, function () { }).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;false&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(undefined, function () { }).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is &#x60;5&#x60;</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(undefined, function () { }).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                    <dt class="error">&#x60;onFulfilled&#x60; is an object</dt>
                    <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                    <dd class="error"><pre><code>resolved(dummy).then(undefined, function () { }).then(function () {
    done();
}, nonFunction);</code></pre></dd>
                    <dd class="error">TypeError: Cannot read property &#x27;then&#x27; of undefined</dd>
                  </dl>
                </section>
              </dl>
            </section>
          </dl>
        </section>
        <section class="suite">
          <h1>2.2.2: If &#x60;onFulfilled&#x60; is a function,</h1>
          <dl>
            <section class="suite">
              <h1>2.2.2.1: it must be called after &#x60;promise&#x60; is fulfilled, with &#x60;promise&#x60;&#x2019;s fulfillment value as its first argument.</h1>
              <dl>
                <dt class="error">already-fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>test(resolved(value), done);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">immediately-fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
d.resolve(value);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">eventually-fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
setTimeout(function () {
    d.resolve(value);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
            <section class="suite">
              <h1>2.2.2.2: it must not be called before &#x60;promise&#x60; is fulfilled</h1>
              <dl>
                <dt class="error">fulfilled after a delay</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var isFulfilled = false;
d.promise.then(function onFulfilled() {
    assert.strictEqual(isFulfilled, true);
    done();
});
setTimeout(function () {
    d.resolve(dummy);
    isFulfilled = true;
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt>never fulfilled</dt>
                <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd><pre><code>var d = deferred();
var onFulfilledCalled = false;
d.promise.then(function onFulfilled() {
    onFulfilledCalled = true;
    done();
});
setTimeout(function () {
    assert.strictEqual(onFulfilledCalled, false);
    done();
}, 150);</code></pre></dd>
              </dl>
            </section>
            <section class="suite">
              <h1>2.2.2.3: it must not be called more than once.</h1>
              <dl>
                <dt class="error">already-fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var timesCalled = 0;
resolved(dummy).then(function onFulfilled() {
    assert.strictEqual(++timesCalled, 1);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">trying to fulfill a pending promise more than once, immediately</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = 0;
d.promise.then(function onFulfilled() {
    assert.strictEqual(++timesCalled, 1);
    done();
});
d.resolve(dummy);
d.resolve(dummy);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">trying to fulfill a pending promise more than once, delayed</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = 0;
d.promise.then(function onFulfilled() {
    assert.strictEqual(++timesCalled, 1);
    done();
});
setTimeout(function () {
    d.resolve(dummy);
    d.resolve(dummy);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">trying to fulfill a pending promise more than once, immediately then delayed</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = 0;
d.promise.then(function onFulfilled() {
    assert.strictEqual(++timesCalled, 1);
    done();
});
d.resolve(dummy);
setTimeout(function () {
    d.resolve(dummy);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when multiple &#x60;then&#x60; calls are made, spaced apart in time</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = [0, 0, 0];
d.promise.then(function onFulfilled() {
    assert.strictEqual(++timesCalled[0], 1);
});
setTimeout(function () {
    d.promise.then(function onFulfilled() {
        assert.strictEqual(++timesCalled[1], 1);
    });
}, 50);
setTimeout(function () {
    d.promise.then(function onFulfilled() {
        assert.strictEqual(++timesCalled[2], 1);
        done();
    });
}, 100);
setTimeout(function () {
    d.resolve(dummy);
}, 150);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when &#x60;then&#x60; is interleaved with fulfillment</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = [0, 0];
d.promise.then(function onFulfilled() {
    assert.strictEqual(++timesCalled[0], 1);
});
d.resolve(dummy);
d.promise.then(function onFulfilled() {
    assert.strictEqual(++timesCalled[1], 1);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
          </dl>
        </section>
        <section class="suite">
          <h1>2.2.3: If &#x60;onRejected&#x60; is a function,</h1>
          <dl>
            <section class="suite">
              <h1>2.2.3.1: it must be called after &#x60;promise&#x60; is rejected, with &#x60;promise&#x60;&#x2019;s rejection reason as its first argument.</h1>
              <dl>
                <dt class="error">already-rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>test(rejected(reason), done);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">immediately-rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
d.reject(reason);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">eventually-rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
setTimeout(function () {
    d.reject(reason);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
            <section class="suite">
              <h1>2.2.3.2: it must not be called before &#x60;promise&#x60; is rejected</h1>
              <dl>
                <dt class="error">rejected after a delay</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var isRejected = false;
d.promise.then(null, function onRejected() {
    assert.strictEqual(isRejected, true);
    done();
});
setTimeout(function () {
    d.reject(dummy);
    isRejected = true;
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt>never rejected</dt>
                <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd><pre><code>var d = deferred();
var onRejectedCalled = false;
d.promise.then(null, function onRejected() {
    onRejectedCalled = true;
    done();
});
setTimeout(function () {
    assert.strictEqual(onRejectedCalled, false);
    done();
}, 150);</code></pre></dd>
              </dl>
            </section>
            <section class="suite">
              <h1>2.2.3.3: it must not be called more than once.</h1>
              <dl>
                <dt class="error">already-rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var timesCalled = 0;
rejected(dummy).then(null, function onRejected() {
    assert.strictEqual(++timesCalled, 1);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">trying to reject a pending promise more than once, immediately</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = 0;
d.promise.then(null, function onRejected() {
    assert.strictEqual(++timesCalled, 1);
    done();
});
d.reject(dummy);
d.reject(dummy);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">trying to reject a pending promise more than once, delayed</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = 0;
d.promise.then(null, function onRejected() {
    assert.strictEqual(++timesCalled, 1);
    done();
});
setTimeout(function () {
    d.reject(dummy);
    d.reject(dummy);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">trying to reject a pending promise more than once, immediately then delayed</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = 0;
d.promise.then(null, function onRejected() {
    assert.strictEqual(++timesCalled, 1);
    done();
});
d.reject(dummy);
setTimeout(function () {
    d.reject(dummy);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when multiple &#x60;then&#x60; calls are made, spaced apart in time</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = [0, 0, 0];
d.promise.then(null, function onRejected() {
    assert.strictEqual(++timesCalled[0], 1);
});
setTimeout(function () {
    d.promise.then(null, function onRejected() {
        assert.strictEqual(++timesCalled[1], 1);
    });
}, 50);
setTimeout(function () {
    d.promise.then(null, function onRejected() {
        assert.strictEqual(++timesCalled[2], 1);
        done();
    });
}, 100);
setTimeout(function () {
    d.reject(dummy);
}, 150);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when &#x60;then&#x60; is interleaved with rejection</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var timesCalled = [0, 0];
d.promise.then(null, function onRejected() {
    assert.strictEqual(++timesCalled[0], 1);
});
d.reject(dummy);
d.promise.then(null, function onRejected() {
    assert.strictEqual(++timesCalled[1], 1);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
          </dl>
        </section>
        <section class="suite">
          <h1>2.2.4: &#x60;onFulfilled&#x60; or &#x60;onRejected&#x60; must not be called until the execution context stack contains only platform code.</h1>
          <dl>
            <section class="suite">
              <h1>&#x60;then&#x60; returns before the promise becomes fulfilled or rejected</h1>
              <dl>
                <dt class="error">already-fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>test(resolved(value), done);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">immediately-fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
d.resolve(value);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">eventually-fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
setTimeout(function () {
    d.resolve(value);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">already-rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>test(rejected(reason), done);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">immediately-rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
d.reject(reason);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">eventually-rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
test(d.promise, done);
setTimeout(function () {
    d.reject(reason);
}, 50);</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
            <section class="suite">
              <h1>Clean-stack execution ordering tests (fulfillment case)</h1>
              <dl>
                <dt>when &#x60;onFulfilled&#x60; is added immediately before the promise is fulfilled</dt>
                <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd><pre><code>var d = deferred();
var onFulfilledCalled = false;
d.promise.then(function onFulfilled() {
    onFulfilledCalled = true;
});
d.resolve(dummy);
assert.strictEqual(onFulfilledCalled, false);</code></pre></dd>
                <dt>when &#x60;onFulfilled&#x60; is added immediately after the promise is fulfilled</dt>
                <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd><pre><code>var d = deferred();
var onFulfilledCalled = false;
d.resolve(dummy);
d.promise.then(function onFulfilled() {
    onFulfilledCalled = true;
});
assert.strictEqual(onFulfilledCalled, false);</code></pre></dd>
                <dt class="error">when one &#x60;onFulfilled&#x60; is added inside another &#x60;onFulfilled&#x60;</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var promise = resolved();
var firstOnFulfilledFinished = false;
promise.then(function () {
    promise.then(function () {
        assert.strictEqual(firstOnFulfilledFinished, true);
        done();
    });
    firstOnFulfilledFinished = true;
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when &#x60;onFulfilled&#x60; is added inside an &#x60;onRejected&#x60;</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var promise = rejected();
var promise2 = resolved();
var firstOnRejectedFinished = false;
promise.then(null, function () {
    promise2.then(function () {
        assert.strictEqual(firstOnRejectedFinished, true);
        done();
    });
    firstOnRejectedFinished = true;
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when the promise is fulfilled asynchronously</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var firstStackFinished = false;
setTimeout(function () {
    d.resolve(dummy);
    firstStackFinished = true;
}, 0);
d.promise.then(function () {
    assert.strictEqual(firstStackFinished, true);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
            <section class="suite">
              <h1>Clean-stack execution ordering tests (rejection case)</h1>
              <dl>
                <dt>when &#x60;onRejected&#x60; is added immediately before the promise is rejected</dt>
                <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd><pre><code>var d = deferred();
var onRejectedCalled = false;
d.promise.then(null, function onRejected() {
    onRejectedCalled = true;
});
d.reject(dummy);
assert.strictEqual(onRejectedCalled, false);</code></pre></dd>
                <dt>when &#x60;onRejected&#x60; is added immediately after the promise is rejected</dt>
                <dt>/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd><pre><code>var d = deferred();
var onRejectedCalled = false;
d.reject(dummy);
d.promise.then(null, function onRejected() {
    onRejectedCalled = true;
});
assert.strictEqual(onRejectedCalled, false);</code></pre></dd>
                <dt class="error">when &#x60;onRejected&#x60; is added inside an &#x60;onFulfilled&#x60;</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var promise = resolved();
var promise2 = rejected();
var firstOnFulfilledFinished = false;
promise.then(function () {
    promise2.then(null, function () {
        assert.strictEqual(firstOnFulfilledFinished, true);
        done();
    });
    firstOnFulfilledFinished = true;
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when one &#x60;onRejected&#x60; is added inside another &#x60;onRejected&#x60;</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var promise = rejected();
var firstOnRejectedFinished = false;
promise.then(null, function () {
    promise.then(null, function () {
        assert.strictEqual(firstOnRejectedFinished, true);
        done();
    });
    firstOnRejectedFinished = true;
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">when the promise is rejected asynchronously</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>var d = deferred();
var firstStackFinished = false;
setTimeout(function () {
    d.reject(dummy);
    firstStackFinished = true;
}, 0);
d.promise.then(null, function () {
    assert.strictEqual(firstStackFinished, true);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
          </dl>
        </section>
        <section class="suite">
          <h1>2.2.5 &#x60;onFulfilled&#x60; and &#x60;onRejected&#x60; must be called as functions (i.e. with no &#x60;this&#x60; value).</h1>
          <dl>
            <section class="suite">
              <h1>strict mode</h1>
              <dl>
                <dt class="error">fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>resolved(dummy).then(function onFulfilled() {
    &#x22;use strict&#x22;;
    assert.strictEqual(this, undefined);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>rejected(dummy).then(null, function onRejected() {
    &#x22;use strict&#x22;;
    assert.strictEqual(this, undefined);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
            <section class="suite">
              <h1>sloppy mode</h1>
              <dl>
                <dt class="error">fulfilled</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>resolved(dummy).then(function onFulfilled() {
    assert.strictEqual(this, global);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
                <dt class="error">rejected</dt>
                <dt class="error">/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js</dt>
                <dd class="error"><pre><code>rejected(dummy).then(null, function onRejected() {
    assert.strictEqual(this, global);
    done();
});</code></pre></dd>
                <dd class="error">Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure &#x22;done()&#x22; is called; if returning a Promise, ensure it resolves. (/Users/zhengqiguang/Work/private/note-book/codes/promise/promise_2.js)</dd>
              </dl>
            </section>
          </dl>
        </section>
        <section class="suite">
          <h1>2.2.6: &#x60;then&#x60; may be called multiple times on the same promise.</h1>
          <dl>
            <section class="suite">
              <h1>2.2.6.1: If/when &#x60;promise&#x60; is fulfilled, all respective &#x60;onFulfilled&#x60; callbacks must execute in the order of their originating calls to &#x60;then&#x60;.</h1>
              <dl>
                <section class="suite">
                  <h1>multiple boring fulfillment handlers</h1>
                  <dl>
